Improve commands to move blob metadata to new table

millerdev opened the pull request:
Add error handling to deal with duplicate metadata states and cases where multiple attachment records point to the same blob. Adding this as part of the `run_sql` command is a little silly as it's beyond simply running SQL, but it was the fastest way I could think of to get it done.

The tests are there because all the different states were getting hard to reason about without them. I plan to remove most of this code after the migration is complete (hopefully very soon).

@nickpell @snopoke

millerdev committed:
Handle duplicate blob keys (simple case)

This is getting a little silly as it's pushing the limits of running SQL, but it was the fastest way I could think of to get it done.

millerdev committed:
Output processing rate

millerdev committed:
Handle more duplicate form attachment meta states

millerdev committed:
Add query to check form attachment migration state

millerdev committed:
Fix tests on sharded envs

millerdev committed:
Refactoring/cleanup

millerdev committed:
lint

millerdev committed:
Fix rate calculation

millerdev committed:
second try at lint

nickpell commented on corehq/form_processor/tests/test_run_sql.py:
```diff
@@ -0,0 +1,271 @@
+        ]
+
+        pre_metas = self.get_metas()
+        print(pformat({get_key(meta, 1) for meta in pre_metas}))  # noqa: E1601
```
For this lint warning I recommend ```from __future__ import print_function``` since ```modernize``` will flag this, despite the ```# noqa: E1601```.

nickpell replied:
(or the change you just pushed 👍 )

nickpell commented on corehq/form_processor/management/commands/run_sql.py:
```diff
@@ -138,7 +138,7 @@ def run(cls, sql, dbname, print_rows):
                 if now > next_update:
                     secs = (now - prev_update).total_seconds() or 1
-                    rate = round(next_processed / float(secs))  # noqa: W1633
+                    rate = int(next_processed / float(secs) + 0.5)
```
clever

millerdev committed:
Merge remote-tracking branch 'origin/master' into dm/blobmeta-move

millerdev committed:
Handle more than two metas reference same key

Also remove superfluous test.

millerdev committed:
Handle blobmeta record with wrong type code

millerdev committed:
Add option to send output to file

millerdev committed:
Make ON CONFLICT clause handle bad type code

millerdev committed:
Make delete command fix bad type code

millerdev committed:
Count deleted items, not updated

millerdev committed:
Fix delete_dup_form_attachments query

The BEGIN statement (now removed) made the first run through ineffective because the transaction was never committed, and therefore all changes were rolled back. Sigh.

millerdev commented:
✅

nickpell reviewed (approved):
I don't have a lot of context, but looks reasonable
